expression/builtin_time.go

```diff
@@ -305,6 +305,32 @@ type builtinTimeDiffSig struct {
 	baseBuiltinFunc
 }
 
+func (b *builtinTimeDiffSig) getStrFsp(strArg string, fsp int) int {
+	if n := strings.IndexByte(strArg, '.'); n >= 0 {
+		lenStrFsp := len(strArg[n+1:])
+		if lenStrFsp <= types.MaxFsp {
+			fsp = int(math.Max(float64(lenStrFsp), float64(fsp)))
+		}
+	}
+	return fsp
+}
+
+func (b *builtinTimeDiffSig) convertArgToTime(sc *variable.StatementContext, arg types.Datum, fsp int) (t types.Time, err error) {
```

Comment:
can we change the definition of `arg` to `Expression`, and check `arg.GetClassType` in switch-case?